File: Server/src/main/java/bgu/spl171/net/impl/TFTP/PacketEncoderDecoder.java
package bgu.spl171.net.impl.TFTP;

import bgu.spl171.net.api.MessageEncoderDecoder;
import bgu.spl171.net.impl.packets.*;

/**
 * Implementation of the MessageEncoderDecoder used to encode and decode data transmitted through TCP/IP communication.
 */
public class PacketEncoderDecoder implements MessageEncoderDecoder<Packet> {

    private LineMessageEncoderDecoder messageEncDec;

    private short opCode = -1;
    private byte[] shortByteArr;
    private int shortByteArrIndex;
    private short dataPacketSize = -1;
    private byte[] dataBytes;
    private int dataBytesIndex;
    private short blockNumber = -1;
    private short errorCode = -1;
    private byte deletedAdded = -1;

    @Override
    public Packet decodeNextByte(byte nextByte) {
        if (opCode == -1) {
            opCode = decodeShortPacketSegment(nextByte, opCode);
            if (opCode == Packet.DIRQ) {
                resetFields();
                return new DIRQPacket();
            }
            if (opCode == Packet.DISC) {
                resetFields();
                return new DISCPacket();
            }
        } else {
            if (isStringMessagePacket())
                return decodeMessagePacket(nextByte);
            if (opCode == Packet.DATA)
                return decodeDATAPacket(nextByte);
            if (opCode == Packet.ACK)
                return decodeACKPack(nextByte);
            if (opCode == Packet.ERROR)
                return decodeERRORPacket(nextByte);
            if (opCode == Packet.BCAST)
                return decodeBCASTPacket(nextByte);
        }
        return null;
    }

    @Override
    public byte[] encode(Packet message) {
        if (message == null)
            return null;
        byte[] res = null;
        byte[] opCodeBytes = encodeOpCode(message);
        if (opCode == Packet.DIRQ || opCode == Packet.DISC) {
            res = opCodeBytes;
        } else if (isStringMessagePacket()) {
            res = encodeMessagePacket(message, opCodeBytes);
        } else if (opCode == Packet.DATA) {
            res = encodeDATAPacket(message, opCodeBytes);
        } else if (opCode == Packet.ACK) {
            res = encodeACKPack(message, opCodeBytes);
        } else if (opCode == Packet.ERROR) {
            res = encodeERRORPacket(message, opCodeBytes);
        } else if (opCode == Packet.BCAST) {
            res = encodeBCASTPacket(message, opCodeBytes);
        }
        resetFields();
        return res;
    }

    private short decodeShortPacketSegment(byte nextByte, short value) {
        if (shortByteArr == null) {
            shortByteArr = new byte[2];
            shortByteArrIndex = 0;
        }
        shortByteArr[shortByteArrIndex++] = nextByte;
        if (shortByteArrIndex == shortByteArr.length) {
            value = bytesToShort(shortByteArr);
            shortByteArrIndex = 0;
        }
        return value;
    }

    private String decodeStringMessage(byte nextByte) {
        if (messageEncDec == null)
            messageEncDec = new LineMessageEncoderDecoder();
        return messageEncDec.decodeNextByte(nextByte);
    }

    private boolean isStringMessagePacket() {
        return (opCode == Packet.RRQ || opCode == Packet.WRQ || opCode == Packet.LOGRQ || opCode == Packet.DELRQ);
    }

    private Packet decodeMessagePacket(byte nextByte) {
        String message = decodeStringMessage(nextByte);
        if (message != null) {
            return newStringMessagePacket(message);
        }
        return null;
    }

    private DATAPacket decodeDATAPacket(byte nextByte) {
        if (dataPacketSize == -1) {
            dataPacketSize = decodeShortPacketSegment(nextByte, dataPacketSize);
        } else {
            if (dataBytes == null) {
                dataBytes = new byte[dataPacketSize];
                dataBytesIndex = 0;
            }
            if (blockNumber == -1) {
                blockNumber = decodeShortPacketSegment(nextByte, blockNumber);
                if (blockNumber != -1 && dataPacketSize == 0) {
                    DATAPacket packet = new DATAPacket(dataPacketSize, blockNumber, dataBytes);
                    resetFields();
                    dataBytes = null;
                    return packet;
                }
            } else {
                dataBytes[dataBytesIndex++] = nextByte;
                if (dataBytesIndex == dataPacketSize) {
                    DATAPacket packet = new DATAPacket(dataPacketSize, blockNumber, dataBytes);
                    resetFields();
                    dataBytes = null;
                    return packet;
                }
            }
        }
        return null;
    }

    private ACKPack decodeACKPack(byte nextByte) {
        blockNumber = decodeShortPacketSegment(nextByte, blockNumber);
        if (blockNumber != -1) {
            ACKPack result = new ACKPack(blockNumber);
            resetFields();
            return result;
        }
        return null;
    }

    private BCASTPacket decodeBCASTPacket(byte nextByte) {
        if (deletedAdded == -1) {
            deletedAdded = nextByte;
        } else {
            String message = decodeStringMessage(nextByte);
            if (message != null) {
                BCASTPacket result = new BCASTPacket(deletedAdded, message);
                resetFields();
                return result;
            }
        }
        return null;
    }

    private ERRORPacket decodeERRORPacket(byte nextByte) {
        if (errorCode == -1) {
            errorCode = decodeShortPacketSegment(nextByte, errorCode);
        } else {
            String message = decodeStringMessage(nextByte);
            if (message != null) {
                ERRORPacket result = new ERRORPacket(errorCode, message);
                resetFields();
                return result;
            }
        }
        return null;
    }

    private Packet newStringMessagePacket(String message) {
        short tempOpCode = opCode;
        resetFields();
        switch (tempOpCode) {
            case Packet.RRQ:
                return new RRQPacket(message);
            case Packet.WRQ:
                return new WRQPacket(message);
            case Packet.LOGRQ:
                return new LOGRQPacket(message);
            case Packet.DELRQ:
                return new DELRQPacket(message);
        }
        return null;
    }

    private String getStrData(Packet message) {
        switch (message.getOpCode()) {
            case Packet.RRQ:
                return ((RRQPacket)message).getFileName();
            case Packet.WRQ:
                return ((WRQPacket)message).getFileName();
            case Packet.LOGRQ:
                return ((LOGRQPacket)message).getUserName();
            case Packet.DELRQ:
                return ((DELRQPacket)message).getFileName();
            case Packet.ERROR:
                return ((ERRORPacket)message).getErrMsg();
            case Packet.BCAST:
                return ((BCASTPacket)message).getFileName();
        }
        return null;
    }

    private byte[] encodeOpCode(Packet message) {
        opCode = message.getOpCode();
        return shortToBytes(opCode);
    }

    private byte[] encodeStrData(Packet message) {
        String strData = getStrData(message);
        if (messageEncDec == null)
            messageEncDec = new LineMessageEncoderDecoder();
        return messageEncDec.encode(strData);
    }

    private byte[] encodePacketSize(Packet message) {
        dataPacketSize = ((DATAPacket)message).getPacketSize();
        return shortToBytes(dataPacketSize);
    }

    private byte[] encodeBlockNumber(Packet message) {
        blockNumber = opCode == Packet.DATA ? ((DATAPacket)message).getBlockNum()
                : ((ACKPack)message).getBlockNum();
        return shortToBytes(blockNumber);
    }

    private byte[] encodeErrorCode(Packet message) {
        errorCode = ((ERRORPacket)message).getErrorCode();
        return shortToBytes(errorCode);
    }

    private byte[] encodeMessagePacket(Packet message, byte[] opCodeBytes) {
        byte[] strDataBytes = encodeStrData(message);
        byte[] res = new byte[strDataBytes.length + 2];
        System.arraycopy(opCodeBytes, 0, res, 0, opCodeBytes.length);
        System.arraycopy(strDataBytes, 0, res, 2, strDataBytes.length);
        return res;
    }

    private byte[] encodeDATAPacket(Packet message, byte[] opCodeBytes) {
        byte[] packetSizeBytes = encodePacketSize(message);
        byte[] blockNumberBytes = encodeBlockNumber(message);
        dataBytes = ((DATAPacket)message).getData();
        byte[] res = new byte[dataBytes.length + 6];
        System.arraycopy(opCodeBytes, 0, res, 0, opCodeBytes.length);
        System.arraycopy(packetSizeBytes, 0, res, 2, packetSizeBytes.length);
        System.arraycopy(blockNumberBytes, 0, res, 4, blockNumberBytes.length);
        System.arraycopy(dataBytes, 0, res, 6, dataBytes.length);
        return res;
    }

    private byte[] encodeACKPack(Packet message, byte[] opCodeBytes) {
        byte[] res = new byte[4];
        byte[] blockNumberBytes = encodeBlockNumber(message);
        System.arraycopy(opCodeBytes, 0, res, 0, opCodeBytes.length);
        System.arraycopy(blockNumberBytes, 0, res, 2, blockNumberBytes.length);
        return res;
    }
    private byte[] encodeERRORPacket(Packet message, byte[] opCodeBytes) {
        byte[] errorCodeBytes = encodeErrorCode(message);
        byte[] strDataBytes = encodeStrData(message);
        byte[] res = new byte[strDataBytes.length + 4];
        System.arraycopy(opCodeBytes, 0, res, 0, opCodeBytes.length);
        System.arraycopy(errorCodeBytes, 0, res, 2, errorCodeBytes.length);
        System.arraycopy(strDataBytes, 0, res, 4, strDataBytes.length);
        return res;
    }

    private byte[] encodeBCASTPacket(Packet message, byte[] opCodeBytes) {
        byte[] strDataBytes = encodeStrData(message);
        byte[] res = new byte[strDataBytes.length + 3];
        System.arraycopy(opCodeBytes, 0, res, 0, opCodeBytes.length);
        res[2] = ((BCASTPacket)message).getDelOrAdd();
        System.arraycopy(strDataBytes, 0, res, 3, strDataBytes.length);
        return res;
    }

    private void resetFields() {
        opCode = -1;
        dataPacketSize = -1;
        blockNumber = -1;
        errorCode = -1;
        deletedAdded = -1;
        dataBytes = null;
        shortByteArr = null;
        shortByteArrIndex = 0;
        dataBytesIndex = 0;
    }

    private short bytesToShort(byte[] byteArr) {
        short result = (short)((byteArr[0] & 0xff) << 8);
        result += (short)(byteArr[1] & 0xff);
        return result;
    }

    private byte[] shortToBytes(short num) {
        byte[] bytesArr = new byte[2];
        bytesArr[0] = (byte)((num >> 8) & 0xFF);
        bytesArr[1] = (byte)(num & 0xFF);
        return bytesArr;
    }


}

File: Client/src/PacketEncoderDecoder.cpp
#include "PacketEncoderDecoder.h"
#include <iterator>
#include <iostream>

PacketEncoderDecoder::PacketEncoderDecoder() : opCode_(-1), packetSize_(-1), blockNum_(-1), errorCode_(-1), data_(), delOrAdd_(-1), shortByteArr_(), message() {
}


PacketEncoderDecoder::~PacketEncoderDecoder() {
}

std::vector<char> PacketEncoderDecoder::encode(const Packet& packet) {
	opCode_ = packet.getOpcode();
	std::vector<char> result(shortToBytes(opCode_));
	
	if (opCode_ == Rrq) {
		encodeRRQPacket(packet, result);
	} else if (opCode_ == Wrq) {
		encodeWRQPacket(packet, result);
	} else if (opCode_ == Data) {
		encodeDATAPacket(packet, result);
	} else if (opCode_ == Ack) {
		encodeACKPacket(packet, result);
	} else if (opCode_ == Error) {
		encodeERRORPacket(packet, result);
	} else if (opCode_ == Logrq) {
		encodeLOGRQPacket(packet, result);
	} else if (opCode_ == Delrq) {
		encodeDELRQPacket(packet, result);
	}
	resetFields();
	return result;
}

void PacketEncoderDecoder::encodeRRQPacket(const Packet& packet, std::vector<char>& result) const {
	const RRQPacket& rrqPacket(dynamic_cast<const RRQPacket&>(packet));
	std::string fileName = rrqPacket.getFileName();
	std::copy(fileName.begin(), fileName.end(), std::back_inserter(result));
	result.push_back('\0');
}

void PacketEncoderDecoder::encodeWRQPacket(const Packet& packet, std::vector<char>& result) const {
	const WRQPacket& wrqPacket(dynamic_cast<const WRQPacket&>(packet));
	std::string fileName = wrqPacket.getFileName();
	std::copy(fileName.begin(), fileName.end(), std::back_inserter(result));
	result.push_back('\0');
}

void PacketEncoderDecoder::encodeDATAPacket(const Packet& packet, std::vector<char>& result) {
	const DATAPacket& dataPacket(dynamic_cast<const DATAPacket&>(packet));
	packetSize_ = dataPacket.getPacketSize();
	blockNum_ = dataPacket.getBlockNum();
	data_ = dataPacket.getData();
	std::vector<char> packetSizeBytes = shortToBytes(packetSize_);
	std::vector<char> blockNumBytes = shortToBytes(blockNum_);
	result.insert(result.end(), packetSizeBytes.begin(), packetSizeBytes.end());
	result.insert(result.end(), blockNumBytes.begin(), blockNumBytes.end());
	result.insert(result.end(), data_.begin(), data_.end());
	data_.clear();
}

void PacketEncoderDecoder::encodeACKPacket(const Packet& packet, std::vector<char>& result) {
	const ACKPacket& ackPacket(dynamic_cast<const ACKPacket&>(packet));
	blockNum_ = ackPacket.getBlockNum();
	std::vector<char> blockNumBytes = shortToBytes(blockNum_);
	result.insert(result.end(), blockNumBytes.begin(), blockNumBytes.end());
}

void PacketEncoderDecoder::encodeERRORPacket(const Packet& packet, std::vector<char>& result) {
	const ERRORPacket& errorPacket(dynamic_cast<const ERRORPacket&>(packet));
	errorCode_ = errorPacket.getErrorCode();
	std::vector<char> errorCodeBytes = shortToBytes(errorCode_);
	std::string errorMessage = errorPacket.getErrorMessage();
	result.insert(result.end(), errorCodeBytes.begin(), errorCodeBytes.end());
	std::copy(errorMessage.begin(), errorMessage.end(), std::back_inserter(result));
	result.push_back('\0');
}

void PacketEncoderDecoder::encodeLOGRQPacket(const Packet& packet, std::vector<char>& result) const {
	const LOGRQPacket& logrqPacket(dynamic_cast<const LOGRQPacket&>(packet));
	std::string userName = logrqPacket.getUserName();
	std::copy(userName.begin(), userName.end(), std::back_inserter(result));
	result.push_back('\0');
}

void PacketEncoderDecoder::encodeDELRQPacket(const Packet& packet, std::vector<char>& result) const {
	const DELRQPacket& delrqPacket(dynamic_cast<const DELRQPacket&>(packet));
	std::string fileName = delrqPacket.getFileName();
	std::copy(fileName.begin(), fileName.end(), std::back_inserter(result));
	result.push_back('\0');
}

Packet* PacketEncoderDecoder::decodeFromKeyboardInput(const std::string& input) {
	auto spaceIndex = input.find_first_of(' ');
	if (spaceIndex == std::string::npos) {
		if (input == "DIRQ") {
			return new DIRQPacket();
		}
		if (input == "DISC") {
			return new DISCPacket();
		}
	} else {
		std::string type = input.substr(0, spaceIndex);
		std::string argument = input.substr(spaceIndex + 1, std::string::npos);
		if (type == "RRQ") {
			return new RRQPacket(argument);
		}
		if (type == "WRQ") {
			return new WRQPacket(argument);
		}
		if (type == "LOGRQ") {
			return new LOGRQPacket(argument);
		}
		if (type == "DELRQ") {
			return new DELRQPacket(argument);
		}
	}
	return nullptr;
}

Packet* PacketEncoderDecoder::decodeNextByte(char nextByte) {
	if (opCode_ == -1) {
		opCode_ = decodeShortPacketSegment(nextByte, opCode_);
	} else {
		if (opCode_ == Data) {
			return decodeDATAPacket(nextByte);
		}
		if (opCode_ == Ack) {
			return decodeACKPacket(nextByte);
		}
		if (opCode_ == Error) {
			return decodeERRORPacket(nextByte);
		}
		if (opCode_ == Bcast) {
			return decodeBCASTPacket(nextByte);
		}
	}
	return nullptr;
}

short PacketEncoderDecoder::decodeShortPacketSegment(short nextByte, short value) {
	if (shortByteArr_.size() < 2) {
		shortByteArr_.push_back(nextByte);
		if (shortByteArr_.size() == 2) {
			value = bytesToShort(shortByteArr_);
			shortByteArr_.clear();
		}
	}
	return value;
}

std::string PacketEncoderDecoder::decodeStringMessage(char nextByte) {
	std::string result;
	data_.push_back(nextByte);
	if (nextByte == '\0') {
		result = std::string(data_.begin(), data_.end());
		data_.clear();
	}
	return result;
}

DATAPacket* PacketEncoderDecoder::decodeDATAPacket(char nextByte) {
	if (packetSize_ == -1) {
		packetSize_ = decodeShortPacketSegment(nextByte, packetSize_);
	} else if (blockNum_ == -1) {
		blockNum_ = decodeShortPacketSegment(nextByte, blockNum_);
		if (blockNum_ != -1 && packetSize_ == 0) {
			DATAPacket* packet = new DATAPacket(packetSize_, blockNum_, data_);
			resetFields();
			return packet;
		}
	} else {
		data_.push_back(nextByte);
		if (data_.size() == packetSize_) {
			DATAPacket* packet = new DATAPacket(packetSize_, blockNum_, data_);
			resetFields();
			return packet;
		}
	}
	return nullptr;
}

ACKPacket* PacketEncoderDecoder::decodeACKPacket(char nextByte) {
	blockNum_ = decodeShortPacketSegment(nextByte, blockNum_);
	if (blockNum_ != -1) {
		ACKPacket* result = new ACKPacket(blockNum_);
		resetFields();
		return result;
	}
	return nullptr;
}

ERRORPacket* PacketEncoderDecoder::decodeERRORPacket(char nextByte) {
	if (errorCode_ == -1) {
		errorCode_ = decodeShortPacketSegment(nextByte, errorCode_);
	} else {
		std::string message = decodeStringMessage(nextByte);
		if (!message.empty()) {
			ERRORPacket* result = new ERRORPacket(errorCode_, message);
			resetFields();
			return result;
		}
	}
	return nullptr;
}

BCASTPacket* PacketEncoderDecoder::decodeBCASTPacket(char nextByte) {
	if (delOrAdd_ == -1) {
		delOrAdd_ = nextByte;
	} else {
		std::string message = decodeStringMessage(nextByte);
		if (!message.empty()) {
			BCASTPacket* result = new BCASTPacket(delOrAdd_, message);
			resetFields();
			return result;
		}
	}
	return nullptr;
}

void PacketEncoderDecoder::resetFields() {
	opCode_ = -1;
	packetSize_ = -1;
	blockNum_ = -1;
	errorCode_ = -1;
	delOrAdd_ = -1;
	data_.clear();
	shortByteArr_.clear();
}


short PacketEncoderDecoder::bytesToShort(const std::vector<char>& bytesArr) {
	short result = (bytesArr[0] & 0xff) << 8;
	result += bytesArr[1] & 0xff;
	return result;
}

std::vector<char> PacketEncoderDecoder::shortToBytes(short num) {
	std::vector<char> bytes;
	bytes.push_back((num >> 8) & 0xFF);
	bytes.push_back(num & 0xFF);
	return bytes;
}
